### PR TITLE
APIs

### PR DESCRIPTION
**Pull request by AliceVlasov**
Allow logged-in users to write, edit, delete, and add public location tags to their comments using an interactive map.

While logged out, the client may only read and interact with the website's content but is not permitted to edit, add, remove, or change any content. If the client logs in using the login button on the main page, the welcome message will be customized with their nickname and they will be able to write comments as well as edit and delete their own comments using the icons that will only appear on their comments, marked with "(Me)" next to the comment's author's name. 

In order to set their nickname, the client must submit a comment which will update their nickname on all of their comments as well as on the welcome page (once the page refreshes). Until this is done, their username default will be set to "New User", and while logged out, "Stranger".

The purpose of the map is 1) to display the markers associated with the visible comments (but not the comments that are not currently visible), 2) allow the user to choose a location to tag to their comment, and 3) find the locations tagged to the visible comments. 

The map only allows the client to choose one location per comment but, until the client clicks submit, they can drag(click and drag), move(click in a new location), and even remove (click on the temporary marker) the marker as they like. The client can also submit the comment without choosing a location, and the comment will then be assigned a default marker location.

Once the form is submitted, the new marker becomes unmovable and the comment appears with a location tag on the top-right. If any client clicks on the location tag, the comment is highlighted and the map pans to the marker linked to that comment. Similarly, if the client clicks on a marker on the map, the comment linked to that marker will be highlighted. If the client deletes a comment, the comment element and it's marker will be deleted permanently. While editing a comment, the client can move the marker once again before resubmitting the comment.

In addition, changing the number of visible comments will also change the number of markers accordingly, and whenever a temporary marker is moved, the map automatically pans to center the marker once again.

Below are screenshots of how the welcome page and comments page look when the client is:
A) logged out
LoggedOutWelcome
LoggedOutComments

B) logged in
LoggedInWelcome
LoggedInEditing

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/MarkerServlet.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    datastore.put(markerEntity);
+
+    long markerId = markerEntity.getKey().getId();
+    response.getWriter().println(markerId);
+  }
+
+    @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // only uncomment if something goes wrong with markers
```
Comment (Felomeng) [inline portfolio/src/main/java/com/google/sps/servlets/MarkerServlet.java:56]: why not delete the code?

File: portfolio/src/main/webapp/script.js [hunk 2]
```diff
@@ -260,21 +298,55 @@ function makeCommentAuthorElement(author) {
...
-    fetch('new-comment', {method: 'POST', body: params}).then(refreshComments);
+
+  //TODO: replace loading map and comments with "loading..."
+  if (!validComment(commentText)) {
+    alert("invalid comment");
+    return;
   }
+
+  const latLng = tempMarker? tempMarker.getPosition(): DEFAULT_COORDS;
+  console.log(latLng.toString());
```
Comment (Felomeng) [inline portfolio/src/main/webapp/script.js:315]: console.log should only be used while debugging.

File: portfolio/src/main/webapp/script.js [hunk 3]
```diff
@@ -314,12 +386,142 @@ function clearComments() {
...
+      {center: DEFAULT_COORDS, zoom: 8});     
+  setMapClickEvents();
+}
+
+/**
+ * Fetches the marker for this comment and adds it to the map
+ */
+function loadMarker(comment, commentElement) {
+  const markerId = comment.markerId;
+  console.log(markerId);
```
Comment (Felomeng) [inline portfolio/src/main/webapp/script.js:449]: same

File: portfolio/src/main/webapp/script.js [hunk 4]
```diff
@@ -260,21 +298,55 @@ function makeCommentAuthorElement(author) {
...
   }
+
+  const latLng = tempMarker? tempMarker.getPosition(): DEFAULT_COORDS;
+  console.log(latLng.toString());
+
+  const params = new URLSearchParams();
+  params.append('lat', latLng.lat());
+  params.append('lng', latLng.lng());
+  params.append('visible', true);
+  console.log(params);
```
Comment (Felomeng) [inline portfolio/src/main/webapp/script.js:321]: same

File: portfolio/src/main/java/com/google/sps/servlets/MarkerServlet.java (hunk 1 above)
Comment (AliceVlasov) [reply portfolio/src/main/java/com/google/sps/servlets/MarkerServlet.java:56]: Whoops, I meant to delete this code and the console.logs from the script file. I'll remove them in the next commit.

File: portfolio/src/main/webapp/script.js (hunk 2 above)
Comment (AliceVlasov) [reply portfolio/src/main/webapp/script.js]: Yes, I meant to remove them before submitting so I'll delete them in the next commit.

**Pull request merged**